src/tcache.c

```diff
@@ -169,6 +167,137 @@ tcache_gc_small(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
+	uintptr_t slab_min = (uintptr_t)addr;
+	uintptr_t slab_max = slab_min + bin_infos[szind].slab_size;
+	/* The exp_grow->next starts from HUGEPAGE. */
+	uintptr_t addr_min = slab_min > HUGEPAGE ? (slab_min - HUGEPAGE) : 0;
+	uintptr_t addr_max = slab_min < ((uintptr_t)(-1) - HUGEPAGE) ?
+	    (slab_min + HUGEPAGE) : (uintptr_t)(-1);
+	cache_bin_sz_t nremote = 0, nremote_hard = 0;
+	void **head = cache_bin->stack_head;
+	/* Scan the entire bin to count the number of remote pointers. */
+	for (void **cur = head; cur < head + ncached; cur++) {
+		assert(*cur != NULL);
+		uintptr_t addr_cur = (uintptr_t)*cur;
+		nremote += (cache_bin_sz_t)(addr_cur < addr_min ||
+		    addr_cur >= addr_max);
+		nremote_hard += (cache_bin_sz_t)(addr_cur < slab_min ||
+		    addr_cur >= slab_max);
+	}
+	assert(nremote_hard >= nremote && nremote_hard <= ncached);
+	/*
+	 * The number of ptrs outside of [addr_min, addr_max) is less than
+	 * the intended flush amount, shrink the range to [slab_min, slab_max).
+	 */
+	if (nremote < nflush) {
+		nremote = nremote_hard;
+		addr_min = slab_min;
+		addr_max = slab_max;
+	}
+	/*
+	 * Update the nflush to the larger value between the intended flush count
+	 * and the number of remote ptrs.
+	 */
+	nflush = nremote > nflush ? nremote : nflush;
+	assert(nflush < ncached || nremote == ncached);
+	if (nremote == 0 || nremote == ncached)	goto label_flush;
+
+	/*
+	 * Move the remote points to the bottom of the bin for flushing.
+	 * As long as moved to the bottom, the order of these nremote ptrs
+	 * does not matter, since they are going to be flushed anyway.
+	 * The rest of the ptrs are moved to the top of the bin, and their
+	 * relative order is not maintained.
+	 */
+	void **swap = NULL;
+	cache_bin_sz_t ntop = ncached - nremote, cnt = 0;
+	assert(ntop > 0 && ntop < ncached);
+	/*
+	 * Scan the [head, head + ntop) part of the cache bin, during which
+	 * bubbling the non-remote ptrs to the top of the bin.
+	 * After this, the [head, head + cnt) part of the bin contains only
+	 * non-remote ptrs, and they're in the same relative order as before.
+	 * While the [head + cnt, head + ntop) part contains only remote ptrs.
+	 */
+	for (void **cur = head; cur < head + ntop; cur++) {
+		assert(*cur != NULL);
+		bool not_remote = ((uintptr_t)*cur >= addr_min &&
+		    (uintptr_t)*cur < addr_max);
+		if (not_remote) {
+			/* Tracks the number of non-remote ptrs seen so far. */
+			cnt++;
+			/*
+			 * There is remote ptr before the current non-remote ptr,
+			 * swap the current non-remote ptr with the remote ptr,
+			 * and increment the swap pointer so that it's still
+			 * pointing to the top remote ptr in the bin.
+			 */
+			if (swap != NULL) {
+				assert(swap < cur);
+				assert((uintptr_t)*swap < addr_min ||
+				    (uintptr_t)*swap >= addr_max);
+				void *tmp = *cur;
+				*cur = *swap;
+				*swap = tmp;
+				swap++;
+				assert(swap <= cur);
+				assert((uintptr_t)*swap < addr_min ||
```

Comment:
I think I get the logic, but it's not easy to follow. To double check: the current approach is trying to maintain the same relative order for the remaining items, is that right? If we don't care about that goal, it can be simplified such as a single loop moving from both ends?